### Reduction operation: the sum of the numbers in the range [0, a)

FUNCIÓN EJEMPLO

In [7]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.038762569427490234 seconds
38.6 ms ± 364 µs per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



3.2 A)

In [8]:
import time

# Crear la lista de números del rango [0, 10**6)
value = 10**6
numeros = list(range(value))

# Método a) Usando un bucle for
initialTime_for = time.time()
suma_for = 0
for numero in numeros:
    suma_for += numero
finalTime_for = time.time()

# Método b) Usando la función sum
initialTime_sum = time.time()
suma_sum = sum(numeros)
finalTime_sum = time.time()

# Mostrar resultados
print("Método a) Usando bucle for:")
print(f"Suma: {suma_for}, Tiempo tomado: {finalTime_for - initialTime_for} segundos\n")

print("Método b) Usando función sum:")
print(f"Suma: {suma_sum}, Tiempo tomado: {finalTime_sum - initialTime_sum} segundos\n")


# Usar %timeit para medir tiempos en IPython (si estás en un entorno compatible)
try:
    get_ipython().run_line_magic('timeit', '-r 2 sum(numeros)')
except NameError:
    print("Las operaciones mágicas de %timeit sólo funcionan en entornos IPython.")

Método a) Usando bucle for:
Suma: 499999500000, Tiempo tomado: 0.07643795013427734 segundos

Método b) Usando función sum:
Suma: 499999500000, Tiempo tomado: 0.005727529525756836 segundos

5.44 ms ± 34.9 µs per loop (mean ± std. dev. of 2 runs, 100 loops each)


3.2 B)

In [9]:
import time
import numpy as np

def sumar_con_bucle_for(array):
    """Sumar los elementos de un array utilizando un bucle for."""
    suma = 0
    for elemento in array:
        suma += elemento
    return suma

def sumar_con_numpy_sum(array):
    """Sumar los elementos de un array utilizando numpy.sum()"""
    return np.sum(array)

def medir_tiempo(funcion, *args):
    """Medir el tiempo que tarda una función en ejecutarse."""
    start_time = time.time()
    resultado = funcion(*args)
    end_time = time.time()
    tiempo = end_time - start_time
    return resultado, tiempo

def main():
    value = 10**6

    # Crear la lista y convertirla a array
    lista = list(range(value))
    array = np.array(lista)

    # Sumar los elementos usando un bucle for
    suma_for, tiempo_for = medir_tiempo(sumar_con_bucle_for, array)
    print(f"Tiempo con bucle for (array de NumPy): {tiempo_for:.6f} segundos")

    # Sumar los elementos usando numpy.sum()
    suma_numpy, tiempo_numpy = medir_tiempo(sumar_con_numpy_sum, array)
    print(f"Tiempo con numpy.sum(): {tiempo_numpy:.6f} segundos")

    # Verificar que ambos resultados son iguales
    assert suma_for == suma_numpy, "Los resultados de las sumas no coinciden"
    print(f"\nSuma total: {suma_numpy}")

if __name__ == "__main__":
    main()

Tiempo con bucle for (array de NumPy): 0.058818 segundos
Tiempo con numpy.sum(): 0.000484 segundos

Suma total: 499999500000


### 3.2 C) Resultados y conclusiones

Los experimentos realizados permiten comparar el rendimiento de la operación de reducción utilizando tres enfoques: el código original en Python, listas nativas y estructuras de NumPy.

---

### 1. Código original

- Tiempo secuencial: `0.03876 s`  
- `%timeit`: `38.6 ms ± 364 µs`  

El comportamiento es el esperado para un bucle en Python: correcto, pero limitado por la interpretación constante del código dentro del bucle.

---

### 2. Uso de listas

**a) Bucle for sobre una lista**

- Tiempo: `0.07643 s`  

Este método es más lento que el código original. Aunque la lista es una estructura nativa de Python, el coste de iterar elemento a elemento sigue siendo elevado.

**b) Función `sum(lista)`**

- Tiempo: `0.00572 s`  
- `%timeit`: `5.44 ms ± 34.9 µs`  

El rendimiento mejora de forma muy significativa. Esto se debe a que `sum()` está implementado internamente en C, lo que permite realizar la reducción sin el overhead del bucle interpretado.

---

### 3. Uso de NumPy

**a) Bucle for sobre un array NumPy**

- Tiempo: `0.058818 s`  

El uso de NumPy no aporta ventaja si la operación consiste en iterar manualmente, porque el bucle sigue siendo Python puro.

**b) `numpy.sum(array)`**

- Tiempo: `0.000484 s`  

Este es el método más rápido con diferencia. NumPy ejecuta la reducción mediante código optimizado en C y emplea técnicas de vectorización que aprovechan mejor el hardware.

---

### Conclusión general

Los resultados muestran que el rendimiento varía de forma notable según el método empleado. Los bucles en Python, ya sea sobre listas o arrays, son los menos eficientes debido al coste de la iteración. La función `sum(lista)` ofrece una mejora importante, pero el mayor rendimiento se obtiene utilizando `numpy.sum`, que realiza la operación de manera interna y optimizada. En tareas de cálculo científico, optar por estructuras vectorizadas y operaciones de bajo nivel marca una diferencia considerable en el tiempo de ejecución.


3.3 A)

In [13]:
import time
import numpy as np
from numba import njit

@njit
def sumar_con_bucle_for(array):
    """Sumar los elementos de un array utilizando un bucle for."""
    suma = 0
    for elemento in array:
        suma += elemento
    return suma

@njit
def sumar_con_numpy_sum(array):
    """Sumar los elementos de un array utilizando numpy.sum()"""
    return np.sum(array)

def medir_tiempo(funcion, *args):
    """Medir el tiempo que tarda una función en ejecutarse."""
    start_time = time.time()
    resultado = funcion(*args)
    end_time = time.time()
    tiempo = end_time - start_time
    return resultado, tiempo

def main():
    value = 10**6

    # Crear la lista y convertirla a array
    lista = list(range(value))
    array = np.array(lista)

    # Sumar los elementos usando un bucle for
    suma_for, tiempo_for = medir_tiempo(sumar_con_bucle_for, array)
    print(f"Tiempo con bucle for (array de NumPy): {tiempo_for:.6f} segundos")

    # Sumar los elementos usando numpy.sum()
    suma_numpy, tiempo_numpy = medir_tiempo(sumar_con_numpy_sum, array)
    print(f"Tiempo con numpy.sum(): {tiempo_numpy:.6f} segundos")

    # Verificar que ambos resultados son iguales
    assert suma_for == suma_numpy, "Los resultados de las sumas no coinciden"
    print(f"\nSuma total: {suma_numpy}")

if __name__ == "__main__":
    main()

Tiempo con bucle for (array de NumPy): 0.054418 segundos
Tiempo con numpy.sum(): 0.049600 segundos

Suma total: 499999500000


### 3.3 B) Resultados con Numba

En esta parte se han comparado los tiempos de ejecución de la suma utilizando NumPy “a secas” y las mismas funciones decoradas con `@njit` de Numba.

Primero, sin Numba (apartado 3.2 B), los resultados fueron:

- Bucle `for` sobre un array de NumPy: `0.058818 s`
- `numpy.sum(array)`: `0.000484 s`

Después, con Numba (apartado 3.3 A), obtenemos:

- Bucle `for` con `@njit` sobre el array de NumPy: `0.054418 s`
- `numpy.sum(array)` dentro de una función con `@njit`: `0.049600 s`

A partir de estos valores se pueden sacar varias conclusiones:

- El bucle `for` mejora ligeramente con `@njit` (de ~0.0588 s a ~0.0544 s), pero la diferencia no es muy grande.
- En cambio, `numpy.sum` se ejecuta muchísimo más rápido sin Numba (`0.000484 s`) que dentro de una función decorada con `@njit` (`0.049600 s` en este caso).

Esto tiene sentido por varios motivos:

1. `numpy.sum` ya está altamente optimizada en C. Numba no puede “mejorarla” mucho más y, en este ejemplo, incluso añade sobrecarga.
2. Numba es especialmente útil cuando acelera bucles escritos en Python puro. Si el trabajo pesado lo hace una función de alto nivel de NumPy, la ventaja de Numba es mucho menor.
3. En problemas tan sencillos como esta suma, el tiempo de compilación y la gestión interna de Numba pueden ser comparables (o incluso mayores) que el tiempo de la operación original optimizada.

En resumen, Numba puede ayudar a acelerar bucles explícitos en Python, pero no siempre aporta beneficios cuando se combina con funciones ya muy optimizadas de NumPy como `numpy.sum()`.